### ISSUE TITLE
Enable lsp-lens-mode by default

### ISSUE THREAD

**Issue by michaelpj** [emacs-lsp/lsp-mode#2249]
Currently it's not on by default. This makes code lenses somewhat hard to discover: I've had a couple of people report that they don't work, when in fact they just hadn't turned the mode on.

Now, it's perfectly configurable, and anyone doing `customize-group lsp-mode` should find it... but for most other things we have the "more featureful" version as the default, and this is an exception. So I think there's an argument for having it on by default. If nothing else, people are more likely to go looking for an option if they want to turn it off!

**Comment by michaelpj** [comment]
Oh, also the documentation doesn't mention that it's behind an option, this makes it look like it should be on by default: https://emacs-lsp.github.io/lsp-mode/page/main-features/#code-lens

**Comment by ericdallo** [comment]
Yeah, we should update the documentation or enable by default, I don't recall exactly why isn't enabled by default, WDYT @yyoncho ?

**Comment by yyoncho** [comment]
ATM I don't think they are reliable enough to have them on by default.

**Comment by michaelpj** [comment]
What's unreliable about them? They seemed okay from my (limited) usage. 

I don't really have a good suggestion if they're not good enough to have on by default. Maybe just a note in the docs.

**Comment by nbfalcon** [comment]
One issue I have seen is that it can make Emacs quite slow (scrolling) for very large code bases (probably because overlays aren't fast), e.g. when navigating trough Emacs' C source using `ccls`.

**Comment by yyoncho** [comment]
> One issue I have seen is that it can make Emacs quite slow (scrolling) for very large code bases (probably because overlays aren't fast), e.g. when navigating trough Emacs' C source using `ccls`.

Hm, that might be a regression, can you post a perf report? The initial implementation had acceptable performance at least in the projects I tested with.

**Comment by ericdallo** [comment]
Any news here? I use lens with clojure and dart and it works really well

**Comment by yyoncho** [comment]
The main issue that I see ATM is that the lenses does not work well with the default company popup.

1. The fact that company is hiding the overlays results in moving the text up and down.
2. If the lens is rendered after the completion dialog is displayed it breaks the dialog.

The described issues does not happen with the posframe based popups.

**Comment by ericdallo** [comment]
@yyoncho I never faced this issue with doom-emacs and company, do you have a repro?

**Comment by yyoncho** [comment]
@ericdallo perform completion above a lens

**Comment by ericdallo** [comment]
![working-completion-lens](https://user-images.githubusercontent.com/7820865/116884126-c0d19100-abfc-11eb-8b6b-d214aad4c58f.gif)

It seems to work

**Comment by yyoncho** [comment]
https://user-images.githubusercontent.com/13259670/116884543-846b5880-ac2f-11eb-94d7-3320c8fcaa41.mp4

See how the lenses are removed and added after the completion.

**Comment by ericdallo** [comment]
I confirmed the issue, but IMHO this is not a blocker

**Comment by andreyorst** [comment]
I've tried turning on lens feature, and noticed some problems.

First may be a blocker to turn this feature on by default, because when lens are displayed above current line the scrolling with `C-n` is blocked when the lens is at the top of the window. Scrolling upwards doesn't have this issue. Changing the setting so lens are displayed at the end of line migrates the issue, e.g. scrolling works. I didn't tried with `emacs -Q` yet.

I'm also not sure if this is supposed to be this way:

![image](https://user-images.githubusercontent.com/19470159/119253606-b3377780-bbba-11eb-818d-d47f55e51dc8.png)

Another question - are lens supposed to be clickable? I've only tested with clojure-lsp, but clicking on `N references` or on `N tests` does nothinng to me, yet the mouse changes shape like when I'm pointing on a link.

**Comment by andreyorst** [comment]
> See how the lenses are removed and added after the completion.

@yyoncho I think this is due to lens using overlays, and company also using overlays. 

Company hides line numbers under completion popup in the very similar way, so I don't think this is particularly `lsp-mode` issue, but a Emacs limitation. Using something like `company-posframe` or `company-box` prevents this from happening.

**Comment by yyoncho** [comment]
@andreyorst yes, that is true. The point is that company-mode is using the overlay frontend by default and if things don't work well by default we will get dozens of bug reports.

On the other side, @ericdallo created a setting to put the lens at the end of the line, I have to test that but most likely they will work much better with company-mode default overlay. So we may change the default placement. 

@ericdallo - have you tested end of the line placement with lsp-ui-sideline?

**Comment by andreyorst** [comment]
> ericdallo - have you tested end of the line placement with lsp-ui-sideline?

(I'm not ericdallo, but) I did for about a week already, and it works without any noticeable issues.

**Comment by ericdallo** [comment]
I'm using the lens at end of the line for almost a week and didn't notice any issues with sideline-diagnostics yet :)

About the click on lens, yeah, clojure-lsp doesn't implement nothing yet for the click.

**Comment by ericdallo** [comment]
@andreyorst latest lsp-mode supports clicking the lens for clojure-lsp now :)

@yyoncho I'm using lens at end of line without any issues with lsp-ui or anything, also I know a lot of clojure-lsp users are using the same way, do you think we can now enable as default?
We can then find issues we didn't find yet and then re-think the default or just fix the issues, WDYT?

**Comment by andreyorst** [comment]
> @andreyorst latest lsp-mode supports clicking the lens for clojure-lsp now :)

Great! I've also been using lens at the end of line since May, with **literally no issues**. Although I need to re-check if the previously mentioned scrolling issue is still a thing. From https://github.com/emacs-lsp/lsp-mode/issues/2249#issuecomment-846526472:

> when lens are displayed above current line the scrolling with `C-n` is blocked when the lens is at the top of the window. Scrolling upwards with `C-p` doesn't have this issue. Changing the setting so lens are displayed at the end of line migrates the issue, e.g. scrolling works. I didn't tried with `emacs -Q` yet.

If anyone can reproduce this please tell me, as I have a pretty non-standard configuration, so it might as well be a bug in my config. I will eventually try this with `emacs -Q`, but it's a bit complicated atm.

**Comment by ericdallo** [comment]
> when lens are displayed above current line the scrolling with C-n is blocked when the lens is at the top of the window. Scrolling upwards with C-p doesn't have this issue. Changing the setting so lens are displayed at the end of line migrates the issue, e.g. scrolling works. I didn't tried with emacs -Q yet.

> If anyone can reproduce this please tell me, as I have a pretty non-standard configuration, so it might as well be a bug in my config. I will eventually try this with emacs -Q, but it's a bit complicated atm.

Since this is not the default behavior anymore, I'm OK on changing the default even if this bug happens
